apng.cpp

```diff
@@ -0,0 +1,457 @@
+
+    png_set_read_fn(d->png_ptr, d, user_read_data);
+
+    png_read_info(d->png_ptr, d->info_ptr);
+
+    // Apply transformations to image data
+    png_set_expand(d->png_ptr);
+    png_set_strip_16(d->png_ptr);
+    png_set_gray_to_rgb(d->png_ptr);
+    png_set_add_alpha(d->png_ptr, 0xFF, PNG_FILLER_AFTER);
+    png_set_bgr(d->png_ptr);
+
+    png_read_update_info(d->png_ptr, d->info_ptr);
+
+    if(!png_get_valid(d->png_ptr, d->info_ptr, PNG_INFO_acTL))
+        goto error;
+
+    width = png_get_image_width(d->png_ptr, d->info_ptr);
+    height = png_get_image_height(d->png_ptr, d->info_ptr);
+
+    d->prev_frame = (uint8_t *) calloc(4, width * height);
+    if (!d->prev_frame)
+        goto error;
+
+    d->prev_rows = (png_bytepp) calloc(sizeof (png_bytep), height);
+    if (!d->prev_rows)
+        goto error;
+
+    for (int i = 0; i < height; i++) {
+        d->prev_rows[i] = &d->prev_frame[i * 4 * width];
+    }
+
+    return d;
+
+error:
+    apng_decoder_release(d);
+    return NULL;
+}
+
+int apng_decoder_get_width(const apng_decoder d) {
+    return png_get_image_width(d->png_ptr, d->info_ptr);
+}
+
+int apng_decoder_get_height(const apng_decoder d) {
+    return png_get_image_height(d->png_ptr, d->info_ptr);
+}
+
+int apng_decoder_get_num_frames(const apng_decoder d) {
+    return png_get_num_frames(d->png_ptr, d->info_ptr);
+}
+
+int apng_decoder_get_frame_width(const apng_decoder d) {
+    return d->frame_header.width;
+}
+
+int apng_decoder_get_frame_height(const apng_decoder d) {
+    return d->frame_header.height;
+}
+
+int apng_decoder_get_prev_frame_delay_num(const apng_decoder d) {
+    return d->frame_header.delay_num;
+}
+
+int apng_decoder_get_prev_frame_delay_den(const apng_decoder d) {
+    return d->frame_header.delay_den;
+}
+
+void apng_decoder_release(apng_decoder d) {
+    png_destroy_read_struct(&d->png_ptr, &d->info_ptr, NULL);
+    free(d->prev_frame);
+    free(d->prev_rows);
+    delete d;
+}
+
+apng_decoder_frame_state apng_decoder_decode_frame_header(apng_decoder d) {
+    if(setjmp(png_jmpbuf(d->png_ptr))) {
+        return apng_decoder_error;
+    }
+
+    if (d->frame_num >= png_get_num_frames(d->png_ptr, d->info_ptr)) {
+        return apng_decoder_eof;
+    }
+
+    png_read_frame_head(d->png_ptr, d->info_ptr);
+
+    if (png_get_valid(d->png_ptr, d->info_ptr, PNG_INFO_fcTL)) {
+        png_get_next_frame_fcTL(d->png_ptr, d->info_ptr,
+            &d->frame_header.width, &d->frame_header.height,
+            &d->frame_header.x_offset, &d->frame_header.y_offset,
+            &d->frame_header.delay_num, &d->frame_header.delay_den,
+            &d->frame_header.dispose_op, &d->frame_header.blend_op);
+    }
+    // todo: case where first frame has no fcTL and is therefore not the first frame of the image
+
+    return apng_decoder_have_next_frame;
+}
+
+void BlendOver(unsigned char * dst, unsigned int dst_width, unsigned char ** rows_src, unsigned int x, unsigned int y, unsigned int w, unsigned int h)
```

Comment:
Out of curiosity, I wonder if opencv has a vectorized function for this operation?

Reply:
I'm not too familiar with OpenCV, but when I checked the documentation I didn't see anything that does this. The closest thing I could find was addWeighted, but that uses a single alpha value for the entire image, rather than using the alpha value for each pixel. It's entirely possible that such a function exists and I didn't notice it, though.